### ISSUE TITLE
Implement network auth system

### ISSUE THREAD

**Issue by XeonSquared** [ShadowKatStudios/OC-Minitel#5]
I want an easy way to provide user/password auth over streams.
At one point I implemented a system as follows:

1. Client sends auth request
2. Server generates a salt
3. Server concatenates password with salt, hashes (sha256 in this case, not ideal)
4. Server sends salt to client
5. Client concatenates password with salt, hashes
6. Client sends hash(salt, password) to server
7. Server compares both results, if they match, the password is correct.

This has a significant downside in that both sides require plaintext passwords, and in the server's case, stored somewhere.

The alternatives are to use other crypto methods to get a password transmitted securely for the server to hash.

Some interesting things to look into:
- [aeslua](https://github.com/SquidDev-CC/aeslua)
- [Ring-LWE key exchange](http://www.computercraft.info/forums2/index.php?/topic/29034-proof-of-concept-ring-lwe-key-exchange/)

**Comment by skyem123** [comment]
You could store the password as a hash+salt on the server, requiring the client to hash twice?

**Comment by XeonSquared** [comment]
The issue with that is that it makes the hashed password the password, you don't need the actual password.

**Comment by skyem123** [comment]
yes, but it does improve security, as it avoids certain issues with password reuse (if you make sure the salts are different!)

**Comment by MrDetonia** [comment]
It might be an idea to implement a simple version of [Kerberos](http://web.mit.edu/kerberos/) authentication. This gets rid of the issue of hash reuse, and also provides an encryption layer to be used after said authentication.

Note this still potentially leaves applications vulnerable to [pass the hash attacks](https://en.wikipedia.org/wiki/Pass_the_hash), which need to be mitigated by creating new Kerberos sessions.

**Comment by XeonSquared** [comment]
So it looks like having passwords of some description stored on the server is non-optional - you can't use them to generate things if they're hashed and salted without compromising the hash.

Instead, here's what I suggest:  
Have a keyring daemon like ssh-agent that holds the actual passwords in memory. This reads the (encrypted) password db from disk, and provides services for generating salted hashes (ie you can give it a salt and request it to hash(password .. salt)) and setting users/passwords, but not actually get the password back out.

Still looking for other options, of course.
